contracts/MasterChef.sol

```diff
@@ -0,0 +1,294 @@
+        farmInfo.accRewardPerShare = farmInfo.accRewardPerShare.add(tokenReward.mul(1e12).div(lpSupply));
+        farmInfo.lastRewardBlock = block.number < farmInfo.endBlock ? block.number : farmInfo.endBlock;
+    }
+
+    function deposit(
+        uint256 _pid,        
+        bytes32 _trustedBlockhash,
+        bytes memory _rlpEncodedBlockHeader,
+        bytes memory _rlpEncodedReceipt,
+        bytes memory _receiptPath,
+        bytes memory _receiptWitness,
+        uint8 _transferEventIndex,
+        bytes32 _receiptsRoot
+    ) public {
+
+        require(_receiptsRoot != bytes32(0), "Invalid receipts root");
+        require(receipts[_receiptsRoot] == false, "Tx already processed");
+
+        receipts[_receiptsRoot] = true;
+
+        // validations 
+        bool proofVerification = EventProof.proveReceiptInclusion(
+            _trustedBlockhash,
+            _rlpEncodedBlockHeader,
+            _rlpEncodedReceipt,
+            _receiptPath,
+            _receiptWitness
+        );
+
+        require(proofVerification == true, "Merkle Proof verification failed");
+
+        // It is an array consisting of below data points :
+        //      contract address on which event is fired
+        //      Array : 
+        //             unique id for event fired --- hash of event signature
+        //             from  
+        //             to 
+        //      value - transferred value
+        RLPReader.RLPItem[] memory rlpReceiptList = RLPReader.toList(RLPReader.toRlpItem(_rlpEncodedReceipt));
+
+        // Here we get all the events fired in the receipt.
+        RLPReader.RLPItem[] memory rlpEventList =  RLPReader.toList(rlpReceiptList[EVENT_INDEX_IN_RECEIPT]);        
+
+        RLPReader.RLPItem[] memory transferEventList = RLPReader.toList(rlpEventList[_transferEventIndex]);
+
+        RLPReader.RLPItem[] memory transferEventParams = RLPReader.toList(
+            transferEventList[TRANSFER_EVENT_PARAMS_INDEX_IN_RECEIPT]
+        );
+
+        require(
+            address(RLPReader.toUint(transferEventList[0])) == CRYPTION_VALIDATOR_SHARE_PROXY_CONTRACT, 
+            "Stake only using Cryption Network validator"
+        );
+
+        // Transfer event signature
+        require(bytes32(transferEventParams[0].toUint())  == TRANSFER_EVENT_SIG , "Invalid event signature");
+
+        // `to` adddress must be masterchef contract
+        require(
+            address(RLPReader.toUint(transferEventParams[2]))  == address(this), 
+            "Shares must be transferred to masterchef"
+        );
+
+        deposit_internal(
+            // _pid,
+            address(transferEventParams[1].toUint()), // `from` address
+            transferEventList[2].toUint() // Value transferred
+        );
+
+    }
+
+    
+    /**
+     * @notice deposit LP token function for msg.sender
+     * @param _amount the total deposit amount
+     */
+    function deposit_internal(address _user, uint256 _amount) internal {
```

Comment:
_depositInternal()

Reply:
Inspiration from recently writing code in JS :stuck_out_tongue: 
Thanks :smiley: